src/App.tsx

```diff
@@ -1,24 +1,220 @@
+import { reducer } from './reducer';
+import { ReducerType } from './enums/Reducer';
 
 export const App: React.FC = () => {
+  const [filter, setFilter] = useState(Filter.ALL);
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, dispatch] = useReducer(reducer, INITIAL_STATE_TEMPTODO);
+  const [processingIds, setProcessingIds] = useState<number[]>([]);
+  const [isCreated, setIsCreated] = useState(false);
+  const [isError, setIsError] = useState<Error>(Error.RESET);
+  const [isArrow, setIsArrow] = useState(false);
+
+  const isProcessing = useCallback((id: number) => {
+    return !!processingIds.includes(id) || id === 0;
+  }, [processingIds]);
+
+  const onShowArrow = useMemo(() => {
+    return setIsArrow(todos.every(todo => todo.completed));
+  }, [todos]);
+
+  const visibleTodos = useMemo(() => {
+    return filteredTodos(todos, filter);
+  }, [todos, filter]);
+
+  const getTodosFromServer = async () => {
+    try {
+      const todosFromServer = await getTodos(USER_ID);
+
+      setTodos(todosFromServer);
+    } catch {
+      setIsError(Error.UPLOAD);
+    }
+  };
+
+  useEffect(() => {
+    getTodosFromServer();
+  }, []);
+
+  const onAdd = useCallback((todo: Todo) => {
+    setIsCreated(true);
+
+    const toAddTodo = async () => {
+      try {
+        const addedTodo = await addTodo(todo);
+
+        setTodos(prev => [...prev, addedTodo]);
+        dispatch({ type: ReducerType.RESET });
+      } catch {
+        setIsError(Error.ADD);
+      } finally {
+        setIsCreated(false);
+      }
+    };
+
+    toAddTodo();
+  }, [todos]);
+
+  const onDelete = useCallback((todoIdToDelete: number) => {
+    setProcessingIds(prev => [...prev, todoIdToDelete]);
+
+    const toRemoveTodo = async () => {
+      try {
+        await removeTodo(todoIdToDelete);
+
+        await getTodosFromServer();
+      } catch {
+        setIsError(Error.REMOVE);
+      } finally {
+        setProcessingIds(prev => prev.filter(id => id !== todoIdToDelete));
+      }
+    };
+
+    toRemoveTodo();
+  }, [todos]);
+
+  const onUpdate = useCallback((todoID: number, data: Partial<Todo>) => {
+    setProcessingIds(prev => [...prev, todoID]);
+
+    const toUpdateTodo = async () => {
+      try {
+        await updateTodo(todoID, data);
+        setProcessingIds((prev) => {
+          return prev.filter(id => id !== todoID);
+        });
+      } catch {
+        setIsError(Error.UPDATE);
+      } finally {
+        await getTodosFromServer();
+      }
+    };
+
+    toUpdateTodo();
+  }, [todos]);
+
+  const handleSubmit = useCallback((e: React.FormEvent) => {
+    e.preventDefault();
+
+    if (!tempTodo.title.trim()) {
+      return setIsError(Error.TITLE);
+    }
+
+    return onAdd(tempTodo);
+  }, [tempTodo]);
+
+  const onClear = useCallback(() => {
+    visibleTodos
+      .map(todo => (todo.completed ? onDelete(todo.id) : todo));
+  }, [visibleTodos]);
+
+  const handleArrow = useCallback(() => {
```

Comment:
Not the best name. Actually it is something that marks all todos as completed if I'm not mistaken. That's part of business logic what happens if designer decides to change arrow to just button `complete all` for example. You'll need to change all names of your variables